### PR TITLE
[Interpreter] Add a simple convnet unit test

### PR DESCRIPTION
**Pull request by beicy**
a convnet to recognize two classes of images (Line and Cross). Now the accuracy rate is 88.5%

### CODE REVIEW

File: tests/unittests/InterpreterTest.cpp [hunk 1]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
   }
 }
 
+/// Generate images in two classes. A "line" is labeled as 0
+/// and a "cross" is labeled as 1
+static void generateImageData(Tensor &images, Tensor &labels, unsigned numSamples, unsigned isTrain) {
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: isTrain seems to be a boolean.

File: tests/unittests/InterpreterTest.cpp [hunk 2]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  for (size_t i = 0; i < numSamples; i++) {
+    bool isLine = i % 2 == 0;
+    L.at({i, 0}) = isLine ? 0 : 1;
+    size_t target;
+    // to make sure testing set is not totally included in 
+    // training set
+    if (isTrain)  {
+      target = rand() % 25 + 1;
+    }
+    else {
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: please make sure to run clang-format

File: tests/unittests/InterpreterTest.cpp [hunk 3]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  int success = 0;
+  auto SMH = result->getVariable()->getHandle<>();
+  for (size_t i = 0; i < numSamples; i++) {
+    bool isLine = (i % 2 == 0);
+    auto A = SMH.at({i, 0});
+    auto B = SMH.at({i, 1});
+    if ((isLine && A > B) || (!isLine && B > A))
+      success++;
+  }
+  EXPECT_TRUE((float)success / numSamples > 0.85);
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: what's the current reported success rate?

Comment (rdzhabarov) [reply tests/unittests/InterpreterTest.cpp]: might just use, EXPECT_GT

File: tests/unittests/InterpreterTest.cpp [hunk 4]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
...
+    size_t target;
+    // to make sure testing set is not totally included in 
+    // training set
+    if (isTrain)  {
+      target = rand() % 25 + 1;
+    }
+    else {
+      target = rand() % 30 + 1;
+    }
+    for (size_t x = 0; x < 32; x++) {
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: since target value is known you could just do single loop here to fill in the tensor.

File: tests/unittests/InterpreterTest.cpp [hunk 5]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {numSamples, 32, 32, 1}, "input",
+                                   VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *ex = mod.createVariable(ElemKind::IndexTy, {numSamples, 1}, "exp",
+                                VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *CV0 = F->createConv("conv1", input, 1, 3, 1, 0, 1);
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: nit: conv, relu etc. no need for 1. Same for var names.

File: tests/unittests/InterpreterTest.cpp [hunk 6]
```diff
@@ -898,6 +898,90 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto image = images.getHandle<>();
+  images.zero();
+  
+  for (size_t i = 0; i < numSamples; i++) {
+    bool isLine = i % 2 == 0;
+    L.at({i, 0}) = isLine ? 0 : 1;
+    size_t target;
+    // to make sure testing set is not totally included in 
+    // training set
+    if (isTrain)  {
```
Comment (jfix71) [inline tests/unittests/InterpreterTest.cpp]: Why is this necessary? `rand()` will not return the same values across different calls to `generateImageData()`.

File: tests/unittests/InterpreterTest.cpp (hunk 3 above)
Comment (jfix71) [reply tests/unittests/InterpreterTest.cpp]: 85% is a bit close to 88.5%, so I think the test might be a bit fragile based on the random numbers provided during training/testing. Maybe use 80% or something? Have you tried using different random number seeds, or looping on the test many times like we discussed before to make sure the test doesn't fail due to accuracy below 85%?

**Comment by artemrakhov** [comment]
Let's aim for 100% accuracy.

1. I thought more about it, and realized that it's OK if train and test data overlap. Imagine that lines and crosses at bottom part of the image are not in the training data. This means that the net doesn't know anything about bottom part of the image. FC.weights cannot be trained properly for these bottom part's pixels because there's no data. Let's have both train and test data equally distributed.

1. Please take a look at `circle` and `nonLinearClassifier` tests. They generate a lot more training data, than the _batchSize_. At each epoch (iteration), training pass uses _batchSize_ consecutive examples (you can read the `runBatch()` code). This allows the net to converge better. The idea is that you need a lot of data to train, but you don't need to feed it all every time.

1. Apparently, TanH works better than Relu in real life. We've seen this before on Xor test.

1. Stability is very important for such tests. I like to run a test 1000 times and make sure it passes. It's easy to do by running: `tests/interpreterTest --gtest_repeat=1000`

I played around with the code and created [this](https://gist.github.com/artemrakhov/0bcc5e225553e93b03fbd19f3c3c884d). It's by no means ready for review or submission, I just wanted to figure out if the task is really solvable. The code passes 1000/1000 reruns.

**Comment by rdzhabarov** [comment]
I find handy to specify the test of interest to speed things up a bit : ```tests/interpreterTest --gtest_repeat=1000 gtest_filter=*simpleConv*```

Test execution time is also important, make sure that it could be run reasonable time in debug mode (longer execution time makes CI painful as well as local test runs).

**Comment by artemrakhov** [comment]
@rdzhabarov, suppose that some pixel is never a cross point in training data. Then, how can FC.weight-matrix learn that this pixel is equally important as all others?

Simple models like this one are not really able to extrapolate.

Actually, I tried splitting train and test this way:
```
    if (isTrain)
      while (target == 6)
        target = nextRandInt(1, 9);
    else
      target = 6;
```
If I change to `stride=3`, then the test passes 1000/1000. But if excluded target is 2 (instead of 6), then I get very few failures (like 997/1000 oks). Here `stride=3` means that we essentially reduce image to smaller size, smaller number of pixels.

In the end, I feel that it's better to have stable test correctly memorizing everything, rather than very rare flaky failures.

Alternatively, we can switch to pre-seeded number generator (not using nextRandInt), and not have to worry about stability.

File: tests/unittests/InterpreterTest.cpp (hunk 6 above)
Comment (beicy) [reply tests/unittests/InterpreterTest.cpp]: Since the number of iteration and numSamples is large, all possible lines and crosses are train without this condition. This isTrain just want to make sure that we generate some images only used for testing

**Comment by bertmaher** [comment]
Our longest-running interpreter test right now is about 7 seconds (`learnSinus`, based on my last local run).  I think we should not add individual tests that take too much longer than that...

**Comment by rdzhabarov** [comment]
> In the end, I feel that it's better to have stable test correctly memorizing everything, rather than very rare flaky failures

@artemrakhov sounds good.

@beicy 
```
rdzhabarov-mbp:debug_build rdzhabarov$ ctest -j 4
Test project /Users/rdzhabarov/src/debug_build
      Start  9: graphOptzTest
      Start  5: interpreterTest
      Start 12: JITTest
      Start  2: gradCheckTest
 1/14 Test  #9: graphOptzTest ....................   Passed    0.46 sec
      Start  6: operatorTest
 2/14 Test  #6: operatorTest .....................   Passed    2.42 sec
      Start 13: GemmTest
 3/14 Test  #2: gradCheckTest ....................   Passed    2.94 sec
      Start 10: quantizationTest
 4/14 Test #10: quantizationTest .................   Passed    1.21 sec
      Start 11: OCLTest
 5/14 Test #11: OCLTest ..........................   Passed    0.38 sec
      Start  7: graphTest
 6/14 Test #13: GemmTest .........................   Passed    2.29 sec
      Start  8: graphGradTest
 7/14 Test  #7: graphTest ........................   Passed    0.54 sec
      Start  1: tensorsTest
 8/14 Test  #8: graphGradTest ....................   Passed    0.20 sec
      Start  4: basicIRTest
 9/14 Test  #4: basicIRTest ......................   Passed    0.01 sec
      Start  3: IROptTest
10/14 Test  #3: IROptTest ........................   Passed    0.03 sec
      Start 14: memoryAllocatorTest
11/14 Test #14: memoryAllocatorTest ..............   Passed    0.01 sec
12/14 Test  #1: tensorsTest ......................   Passed    0.09 sec
13/14 Test  #5: interpreterTest ..................   Passed    9.48 sec
14/14 Test #12: JITTest ..........................   Passed   10.59 sec

100% tests passed, 0 tests failed out of 14

Total Test time (real) =  10.76 sec
```
This is the time for the whole unit test suite. Agree with @bertmaher, something around ~7-10s. (the whole interpreter test takes 10s now, we do not really want to increase that significantly).

**Comment by rdzhabarov** [comment]
@artemrakhov 
>suppose that some pixel is never a cross point in training data. Then, how can FC.weight-matrix learn that this pixel is equally important as all others?

Can max pooling help with this?

**Comment by beicy** [comment]
@rdzhabarov Decreasing the size of images can reduce the debug-mode runtime. For example, keeping other params the same, using 10*10 instead of 32*32, the runtime of debug mode now is 36s. I continue tuning the parameters to reduce the runtime

**Comment by beicy** [comment]
I submitted a new version. The main changes:
1) images size: from 32*32 -> 8*8, numSamples: from 200->100. This makes the debug-mode run time to about 10s.
2) According to my understanding from the above discussion, the cross point is the key in training. Therefore, when generating the training set, all possible positions of a cross point is trained, but the length of a line or a cross is different now. So far, the training set is different from the testing set.  I run the test for 200 times, and 192/200 passes.

**Comment by nadavrot** [comment]
Let's iterate and make sure that this unit test runs below one second. 
For reference, ConvNetJS[1] leans much more complicated patterns in a very short period of time. 

Could you add a stride to reduce the size of the activations that are fed into the FC?  Could you reduce the number of iterations?


[1] https://cs.stanford.edu/people/karpathy/convnetjs/index.html

**Comment by artemrakhov** [comment]
@rdzhabarov yes, MaxPool can help. Also, choosing larger stride and filter size can help the network to figure out missing examples better.

**Comment by beicy** [comment]
@artemrakhov  Thanks for this explanation.  Yes, the runtime is improved a lot in this way!

File: tests/unittests/InterpreterTest.cpp [hunk 7]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  int success = 0;
+  auto SMH = result->getVariable()->getHandle<>();
+  for (size_t i = 0; i < batchSize; i++) {
+    bool isLine = (i % 2 == 0);
+    auto A = SMH.at({i, 0});
+    auto B = SMH.at({i, 1});
+    if ((isLine && A > B) || (!isLine && B > A))
+      success++;
+  }
+  EXPECT_GT((float)success / batchSize, 0.99);
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: nit: We can simply check that success == batchSize (success needs to be changed to unsigned in this case, to avoid signed to unsigned comparison)

File: tests/unittests/InterpreterTest.cpp [hunk 8]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
   }
 }
 
+/// Generate images in two classes. A "line" is labeled as 0
+/// and a "cross" is labeled as 1
+static void generateImageData(Tensor &images, Tensor &labels, unsigned numSamples, bool isTrain) {
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: nit: numSamples can be infered from Tensors' shape: `numSamples = images.dims()[0]`

**Comment by beicy** [comment]
Now the debug_mode runtime is between 0.8-0.9s. And the code passes 295/300

File: tests/unittests/InterpreterTest.cpp [hunk 9]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+    size_t target, end;
+    target = nextRandInt(2, 5);
+    // to make sure testing set is not totally included in 
+    // training set
+    end = isTrain ? 6 : 7;
+
+    for (size_t x = 0; x < end; x++) {
+      for (size_t y = 0; y < end; y++) {
+        if (isLine) {
+	  // line
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: nit: this line contains mix of tab and spaces. Please run clang-format. We have a script for this in our repo:
`./utils/format.sh`

File: tests/unittests/InterpreterTest.cpp [hunk 10]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+
+  // generate the images used for testing.
+  Tensor testImages(ElemKind::FloatTy, {batchSize, 8, 8, 1});
+  Tensor testLabels(ElemKind::IndexTy, {batchSize, 1});
+  generateImageData(testImages, testLabels, batchSize, false);
+  EE.run({input}, {&testImages});
+  int success = 0;
+  auto SMH = result->getVariable()->getHandle<>();
+  for (size_t i = 0; i < batchSize; i++) {
+    bool isLine = (i % 2 == 0);
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: nit: brackets are redundant

File: tests/unittests/InterpreterTest.cpp (hunk 4 above)
Comment (rdzhabarov) [reply tests/unittests/InterpreterTest.cpp]: @beicy i think this is not yet addressed.

File: tests/unittests/InterpreterTest.cpp [hunk 11]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+          // cross
+          if (x == target || y == target)
+            image.at({i, x, y, 0u}) = 1;
+        }
+      }
+    }
+  }
+}
+
+TEST(Interpreter, simpleConv) {
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: This test needs a better name. simpleConv does not describe the test. 

Please add a comment above the test that describes what it does and what it checks.

File: tests/unittests/InterpreterTest.cpp [hunk 12]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  for (size_t i = 0; i < numSamples; i++) {
+    bool isLine = i % 2 == 0;
+    L.at({i, 0}) = isLine ? 0 : 1;
+    size_t target, end;
+    target = nextRandInt(2, 5);
+    // to make sure testing set is not totally included in 
+    // training set
+    end = isTrain ? 6 : 7;
+
+    for (size_t x = 0; x < end; x++) {
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: nit: I think I've already left this comment, that in fact we only need a single loop.
Target is known, target defines either row for line or row and column for cross.

File: tests/unittests/InterpreterTest.cpp [hunk 13]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto &mod = EE.getModule();
+  Function *F = mod.createFunction("main");
+
+  auto *input = mod.createVariable(ElemKind::FloatTy, {batchSize, 8, 8, 1}, "input",
+                                   VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *ex = mod.createVariable(ElemKind::IndexTy, {batchSize, 1}, "exp",
+                                VisibilityKind::Public, Variable::TrainKind::None);
+
+  auto *CV0 = F->createConv("conv", input, 1, 3, 1, 0, 1);
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: super nit: CV, RL, etc.

File: tests/unittests/InterpreterTest.cpp [hunk 14]
```diff
@@ -898,6 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  // generate the images used for testing.
+  Tensor testImages(ElemKind::FloatTy, {batchSize, 8, 8, 1});
+  Tensor testLabels(ElemKind::IndexTy, {batchSize, 1});
+  generateImageData(testImages, testLabels, batchSize, false);
+  EE.run({input}, {&testImages});
+  int success = 0;
+  auto SMH = result->getVariable()->getHandle<>();
+  for (size_t i = 0; i < batchSize; i++) {
+    bool isLine = (i % 2 == 0);
+    auto A = SMH.at({i, 0});
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: maybe give a descriptive name here, like lineWeight and crossWeight and then
isLine && lineWeights > crossWeight will be much more readable.

**Comment by beicy** [comment]
After discussed with @artemrakhov, now generate training/testing set in the same way to make the test stable.

File: tests/unittests/InterpreterTest.cpp [hunk 15]
```diff
@@ -898,12 +898,92 @@ TEST(Interpreter, nonLinearClassifier) {
...
+    } else {
+      for (size_t pos = 0; pos < 8; pos++) {
+        image.at({i, pos, target, 0u}) = 1;
+        image.at({i, target, pos, 0u}) = 1;
+      }
+    }
+  }
+}
+
+/// Test converlutional layer. Use a simple convnet
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: typo.

Comment (nadavrot) [reply tests/unittests/InterpreterTest.cpp]: Also, period at the end of the sentence.

File: tests/unittests/InterpreterTest.cpp [hunk 16]
```diff
@@ -898,12 +898,92 @@ TEST(Interpreter, nonLinearClassifier) {
...
+
+  Tensor images(ElemKind::FloatTy, {numSamples, 8, 8, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels);
+
+  // Training:
+  EE.runBatch(500, {input, ex}, {&images, &labels});
+  EE.compile(CompilationMode::Infer, F);
+
+  // generate the images used for testing.
```
Comment (nadavrot) [inline tests/unittests/InterpreterTest.cpp]: The sentence should start with a capital letter.

File: tests/unittests/InterpreterTest.cpp [hunk 17]
```diff
@@ -898,12 +898,92 @@ TEST(Interpreter, nonLinearClassifier) {
   }
 }
 
+/// Generate images in two classes. A "line" is labeled as 0
+/// and a "cross" is labeled as 1
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: dot at the end.

File: tests/unittests/InterpreterTest.cpp [hunk 18]
```diff
@@ -898,12 +898,92 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  // Training:
+  EE.runBatch(500, {input, ex}, {&images, &labels});
+  EE.compile(CompilationMode::Infer, F);
+
+  // generate the images used for testing.
+  Tensor testImages(ElemKind::FloatTy, {batchSize, 8, 8, 1});
+  Tensor testLabels(ElemKind::IndexTy, {batchSize, 1});
+  generateImageData(testImages, testLabels);
+  EE.run({input}, {&testImages});
+  unsigned success = 0;
```
Comment (rdzhabarov) [inline tests/unittests/InterpreterTest.cpp]: nit: you do not need success counter now.
EXPECT_TRUE(isLine && lineWeight > crossWeight || !isLine && crossWeight > lineWeight) should be enough.

File: tests/unittests/InterpreterTest.cpp [hunk 19]
```diff
@@ -898,12 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  EE.compile(CompilationMode::Infer, F);
+
+  // Generate the images used for testing.
+  Tensor testImages(ElemKind::FloatTy, {batchSize, 8, 8, 1});
+  Tensor testLabels(ElemKind::IndexTy, {batchSize, 1});
+  generateImageData(testImages, testLabels);
+  EE.run({input}, {&testImages});
+  auto SMH = result->getVariable()->getHandle<>();
+  for (size_t i = 0; i < batchSize; i++) {
+    bool isLine = i % 2 == 0;
```
Comment (artemrakhov) [inline tests/unittests/InterpreterTest.cpp]: nit: let's use `testLabels` here

File: tests/unittests/InterpreterTest.cpp [hunk 20]
```diff
@@ -898,12 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  auto *result = F->createSave("ret", SM);
+  Function *TF = glow::differentiate(F, EE.getConfig());
+  EE.compile(CompilationMode::Train, TF);
+
+  Tensor images(ElemKind::FloatTy, {numSamples, 8, 8, 1});
+  Tensor labels(ElemKind::IndexTy, {numSamples, 1});
+  generateImageData(images, labels);
+
+  // Training:
+  EE.runBatch(500, {input, ex}, {&images, &labels});
```
Comment (bertmaher) [inline tests/unittests/InterpreterTest.cpp]: Nit: `numSamples` here (assuming its not just a coincidence that this `500` is the same as that)

File: tests/unittests/InterpreterTest.cpp [hunk 21]
```diff
@@ -898,12 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+      }
+    }
+  }
+}
+
+/// Test the convolutional layer.
+/// Use a simple convnet to learn two classes of images: Line and Cross.
+TEST(Interpreter, convNetForImageRecognition) {
+  unsigned numSamples = 500;
+  unsigned batchSize = 7;
```
Comment (bertmaher) [inline tests/unittests/InterpreterTest.cpp]: Super-nit: `constexpr unsigned`

File: tests/unittests/InterpreterTest.cpp [hunk 22]
```diff
@@ -898,12 +898,89 @@ TEST(Interpreter, nonLinearClassifier) {
...
+  }
+}
+
 TEST(Interpreter, NotImplementedSave) {
   // Interpreter backend does not support a save method.
   // Exercise it and make sure that it fails.
   ExecutionEngine EE;
   auto &mod = EE.getModule();
-  
+
```
Comment (bertmaher) [inline tests/unittests/InterpreterTest.cpp]: Please revert the unintentional whitespace change.

Comment (rdzhabarov) [reply tests/unittests/InterpreterTest.cpp]: this might be legit after clang-format :P

File: tests/unittests/InterpreterTest.cpp (hunk 20 above)
Comment (beicy) [reply tests/unittests/InterpreterTest.cpp]: This is the number of iteration. Yes, it is just a coincidence that it is the same as the numSamples:).

File: tests/unittests/InterpreterTest.cpp (hunk 22 above)
Comment (beicy) [reply tests/unittests/InterpreterTest.cpp]: Yes, this one comes after I run utils/format.sh. Will revert it.
